akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.fusing
+
+import akka.annotation.InternalApi
+import akka.dispatch.ExecutionContexts
+import akka.stream.{ AbruptStageTerminationException, Attributes, FlowShape, Inlet, NeverMaterializedException, Outlet }
+import akka.stream.scaladsl.{ Flow, Keep, Source }
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+import akka.util.OptionVal
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+
+@InternalApi private[akka] final class FutureFlow[In, Out, M](future: Future[Flow[In, Out, M]])
+    extends GraphStageWithMaterializedValue[FlowShape[In, Out], Future[M]] {
+  val in = Inlet[In](s"${this}.in")
+  val out = Outlet[Out](s"${this}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val innerMatValue = Promise[M]
+    val logic = new GraphStageLogic(shape) {
+
+      //seems like we must set handlers BEFORE preStart
+      setHandlers(in, out, initializing)
+
+      override def preStart(): Unit = {
+        super.preStart()
+        future.value match {
```

Comment:
I like the comment from `LazyFlow` on why this is required, but maybe that's just because I'm not as familiar with the trick.

```
          // check if the stage is still in need for the lazy flow
          // (there could have been an onUpstreamFailure or onDownstreamFinish in the meantime that has completed the promise)
```